Add ability to suppress plugins from AMP responses

westonruter opened the pull request:
## Summary

Fixes #4477.
Fixes #3474.

This PR introduces a new Suppressed Plugins (Plugin Suppression) section in the AMP settings screen which is shown when:

1. There are plugins which have been detected to cause validation errors (in non-stale validated URLs).
2. Such plugins have been suppressed.

![image](https://user-images.githubusercontent.com/134745/85165614-f0f98680-b21a-11ea-95a1-23e499fc16af.png)

<details>
<summary><small>Old screenshot</small></summary>

![image](https://user-images.githubusercontent.com/134745/81755178-eb1ac380-946c-11ea-9fbf-427e4081cbda.png)

</details>

The Suppressed Plugins section is omitted entirely if no validation errors are being caused by plugins.

When a plugin is causing validation errors or it has been suppressed, it will appear in the list of initially-visible plugins with select dropdowns. <!--<del>Other plugins which are not known to be causing validation errors are hidden in a collapsed `details` element.</del>--> The other plugins are hidden, and the entire section will appear only if there are any plugins that have been suppressed or if there are other plugins which are causing validation errors. When there are validation errors associated with a given plugin, the errors will be listed with each plugin to provide you with a quick way to evaluate whether the errors are serious enough to warrant turning off the plugin on AMP responses. Validation errors which are bolded are those which are unreviewed. Validation errors with markup removed have a green bullet, and validation errors with red markup have a red bullet.

![image](https://user-images.githubusercontent.com/134745/85165474-c0195180-b21a-11ea-8f20-3dd3775c8284.png)
<details>
<summary><small>Old screenshot</small></summary>

![image](https://user-images.githubusercontent.com/134745/81755276-32a14f80-946d-11ea-9fb9-d4971ada0a0d.png)

</details>

The validation errors are only accounted for when the recent validated URLs are not stale. In other words, when a plugin has been updated and the validated URLs were last checked with an older version of the plugin, then the plugins won't be listed among those which are candidates for suppression. Similarly, once a plugin has been suppressed and then the user updates the plugin, a message will be added next to the plugin to indicate that a new version has been installed and that it should perhaps be re-checked for validity since the issue may not be present anymore:

![image](https://user-images.githubusercontent.com/134745/85165277-6f095d80-b21a-11ea-8802-8a8242b9dc1a.png)
<details>
<summary><small>Old screenshot</small></summary>

![image](https://user-images.githubusercontent.com/134745/81755357-67ada200-946d-11ea-8a99-d25d61a3eafb.png)

</details>

Plugins are suppressed by:

* Unhooking filters and actions whose callbacks are defined in the plugin. 
* Overriding shortcodes with plugin-defined callbacks to just `__return_empty_string`.
* Overriding widgets with plugin-defined callbacks to just `__return_null`. In addition to this, the display is also blocked with the `widget_display_callback` filter to ensure widgets rendered with `the_widget()` are accounted for.
* Overriding dynamic blocks with plugin-defined callbacks to just `__return_empty_string` and to also nullify the frontend `style` and `script`.

Some plugins to help with testing (which have also been adapted to be included as part of the tests):

* [Bad Shortcode](https://gist.github.com/westonruter/6880918d45deb39c89e17f3b76583521)
* [Bad Widget](https://gist.github.com/westonruter/43c6e09354a4bcac008cc14d248e804c)
* [Bad Block](https://gist.github.com/westonruter/e4331b3f6d03aac075b86b226acd3ba5)
* [Bad Hooks](https://gist.github.com/westonruter/f451337351e803044413d942f307229b)

When all of those plugins are active and a post is validated which has the shortcode, widget, and block all present on its template, the validation errors will show as:

![image](https://user-images.githubusercontent.com/134745/81493031-bed72b00-9251-11ea-89d4-3bbd18c8c53d.png)

When all of these are suppressed, no validation errors should be raised by any of them:

![image](https://user-images.githubusercontent.com/134745/81493046-d3b3be80-9251-11ea-9d62-7eb5cb94873d.png)

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

westonruter committed:
Add suppressed plugins option

westonruter committed:
Keep track of the plugin version which was suppressed

westonruter committed:
Include plugin versions in the validated environment

See #3474 and specifically comment https://github.com/ampproject/amp-wp/issues/3474#issuecomment-623865267

westonruter committed:
Ensure get_plugins() function is loaded

westonruter committed:
Hide non-suppressant plugins and show validation errors

westonruter commented:
This isn't ready for review yet.

westonruter committed:
Store plugin slugs instead of files for suppressed list

westonruter committed:
Add initial plugin suppression logic for hooks

westonruter committed:
Add suppression of plugin shortcodes

westonruter committed:
Suppress plugin blocks

westonruter committed:
Remove redundancy in checking callback sources

westonruter committed:
Suppress plugin widgets

westonruter committed:
Fix test failing with undefined index

westonruter committed:
Ensure plugin suppression only happens on AMP pages

westonruter committed:
Pass suppressed plugins as params

westonruter committed:
Re-check a validated URL after changing suppressed plugins

westonruter commented on src/PluginSuppression.php:
```diff
@@ -0,0 +1,140 @@
+		global $wp_registered_widgets;
+		foreach ( $wp_registered_widgets as $widget_id => &$registered_widget ) {
+			if ( $this->is_callback_plugin_suppressed( $registered_widget['callback'], $suppressed_plugins ) ) {
+				$registered_widget['callback'] = '__return_empty_string';
```
```suggestion
				$registered_widget['callback'] = '__return_null';
```

amedina commented on includes/options/class-amp-options-menu.php:
```diff
@@ -423,6 +434,227 @@ private function list_template_conditional_options( $options, $parent = null ) {
+							echo esc_html(
+								sprintf(
+									/* translators: %s is the count of plugins */
+									_n(
```
Let's find an alternative copy for `Other %d plugins`. When there are no plugins with validation issues `other` does not apply. Perhaps `Compliant plugins` or something like that?

westonruter replied:
Good catch. Addressed now that the Other section is removed entirely in d3e772e.

westonruter committed:
Omit suppressed plugins section if no erroring/suppressed plugins; omit others list

westonruter committed:
Omit error duplicates from suppressed plugin list

westonruter committed:
Override widget callback to return null instead of empty string (since widgets print)

westonruter commented:
I still have to add tests, but this is ready for review otherwise.

pierlon commented on includes/options/class-amp-options-manager.php:
```diff
@@ -272,12 +273,85 @@ public static function validate_options( $new_options ) {
+		if ( ! empty( $deleted_suppressed_plugins ) || ! empty( $inserted_suppressed_plugins ) ) {
+			add_action(
+				'update_option_' . self::OPTION_NAME,
+				function () {
```
```suggestion
				static function () {
```

pierlon commented on includes/options/class-amp-options-menu.php:
```diff
@@ -423,6 +436,191 @@ private function list_template_conditional_options( $options, $parent = null ) {
+					get_option( 'active_plugins', [] )
+				)
+			)
+		);
```
Perhaps this could be refactored so that it could be easier to understand what's going on here.

westonruter replied:
Yes, good. Done in d24b29c6bff37e676196dbe8dd89bad95615e4b9.

pierlon commented on includes/options/class-amp-options-menu.php:
```diff
@@ -423,6 +436,191 @@ private function list_template_conditional_options( $options, $parent = null ) {
+					array_keys( AMP_Options_Manager::get_option( Option::SUPPRESSED_PLUGINS ) )
+				),
+				array_map(
+					function ( $plugin_file ) {
```
```suggestion
					static function ( $plugin_file ) {
```

pierlon commented on includes/validation/class-amp-validated-url-post-type.php:
```diff
@@ -890,7 +953,7 @@ public static function get_validated_environment() {
 	 *     @type string $theme   The theme that was active but is no longer. Absent if theme is the same.
-	 *     @type array  $plugins Plugins that used to be active but are no longer, or which are active now but weren't. Absent if the plugins were the same.
+	 *     @type array  $plugins Plugins that used to be active but are no longer, or which are active now but weren't. Also includes plugins that have version updates. Absent if the plugins were the same.
 	 *     @type array  $options Options that are now different. Absent if the options were the same.
```
Is an array of stale options still supposed to be returned here?

westonruter replied:
Actually, you identified some code that needed to be resurrected to address #3474. Resurrection implemented in 4e1c478.

pierlon commented on includes/validation/class-amp-validated-url-post-type.php:
```diff
@@ -865,16 +865,79 @@ public static function store_validation_errors( $validation_errors, $url, $args
+					continue;
+				}
+				foreach ( $validation_error['data']['sources'] as $source ) {
+					if ( ! isset( $source['type'], $source['name'] ) || ! $validation_error['term'] instanceof WP_Term ) {
```
Does `$validation_error['term']` needs to be checked again if it's an instance of `WP_Term`?

```suggestion
					if ( ! isset( $source['type'], $source['name'] ) {
```

pierlon commented on includes/options/class-amp-options-menu.php:
```diff
@@ -423,6 +436,191 @@ private function list_template_conditional_options( $options, $parent = null ) {
+			<ul>
+				<?php foreach ( $plugins as $plugin_file => $plugin ) : ?>
+					<?php
+					$plugin_slug = strtok( $plugin_file, '/' );
```
What about in (rare) cases where multiple plugins can be found in the same directory?

westonruter replied:
Yeah, it's a good point. The reality is that `AMP_Validation_Manager::get_source()` is not able to detect such nested plugins anyway. The source slug extracted from a plugin will always be just the directory name if the plugin is not a single file:

https://github.com/ampproject/amp-wp/blob/9d662d07da228c6cc8dedfab5abc724feaf39177/includes/validation/class-amp-validation-manager.php#L1636-L1639

So that is why the the directory alone is being considered (or else just the file name if the plugin is not in a directory).

pierlon commented on includes/options/class-amp-options-menu.php:
```diff
@@ -423,6 +436,191 @@ private function list_template_conditional_options( $options, $parent = null ) {
+						</label>
+						<?php if ( $is_suppressed && version_compare( $suppressed_plugins[ $plugin_slug ][ Option::SUPPRESSED_PLUGINS_LAST_VERSION ], $plugins[ $plugin_file ]['Version'], '!=' ) ) : ?>
+							<small>
+								<?php if ( $suppressed_plugins[ $plugin_slug ][ Option::SUPPRESSED_PLUGINS_LAST_VERSION ] && $plugins[ $plugin_file ]['Version'] ) : ?>
```
Should this condition be checked before comparing versions above?

westonruter replied:
This is here in the rare situation where a plugin was updated to add a version which previously lacked it, or it was updated to remove the version where it previously had one.

pierlon commented:
This should also address #3474.

westonruter committed:
Make closures static

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Remove redundant check for whether validation error has WP_Term

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Clarify logic in get_suppressible_plugin_sources

westonruter committed:
Fix syntax error after ccbf771b

westonruter committed:
Include options in validated environment and staleness check

See #3474

westonruter committed:
Fix PHPStan analysis by unsetting WP_Block_Type script and style instead of setting to null

westonruter commented:
Thoughts on the use of strike-through styling when a plugin is checked? I realize that may not be so helpful.

amedina commented:
I personally don't like the strike through approach much. IIUC the goal is to make sure users understand clearly that the given plugin has been disabled on AMP. I would like to see that clearly conveyed. Perhaps a label next to the plugin name?

westonruter commented:
> I personally don't like the strike through approach much. IIUC the goal is to make sure users understand clearly that the given plugin has been disabled on AMP. I would like to see that clearly conveyed. Perhaps a label next to the plugin name?

I think perhaps the strike-through is entirely redundant. The checked state indicates what is suppressed. I'll just remove the styling.

westonruter committed:
Remove superflous line-through styling

pierlon commented:
> I think perhaps the strike-through is entirely redundant. The checked state indicates what is suppressed.

Agreed. I also think the updated suppressed plugins could be more indicative. Maybe having them highlighted, or even put into a separate group, for example.

schlessera commented:
I think the checkboxes might visually indicate the opposite of what they should. Normally, a checked checkbox indicates "enabled", "on", "active", ...

How about using similar comboboxes than with the validation error lists here, to keep the same language? Instead of `Removed`/`Kept`, they could say `Suppressed`/`Active` or so.

westonruter committed:
Display suppressed plugins in a table

westonruter commented:
@schlessera @pierlon What do you think of this:

![image](https://user-images.githubusercontent.com/134745/83346204-4d7a1d80-a2cf-11ea-983e-8087d420a083.png)

westonruter committed:
Include more plugin information (URL and author)

westonruter commented:
I'm not entirely sure of the title of the “Details” column. I added a datetime at which the plugin was suppressed in order to avoid the cell from being blank. I've also added the plugin URL, description and author so that the user can get more context for the plugin to decide whether it is safe to suppress:

![image](https://user-images.githubusercontent.com/134745/83359360-98774d80-a32e-11ea-9423-44e5872cede1.png)

Aside: Someone should really reach out to that Weston Ruter guy to educate him on how to write plugins so they don't cause validation errors!

westonruter committed:
Remove redundant plugin sorting, unused ID

westonruter committed:
Move nested loop into helper method

westonruter committed:
Reuse plugin variable instead of looking up the plugin by key

westonruter committed:
Introduce PluginRegistry class

amedina commented:
I am not too keen on putting the details of the offending plugin functionality on this screen. The amount of real estate is very limited and the details may not convey the right information.

amedina commented:
The description of the `Suppressed plugins` section is too verbose. We should simplify it and link to associated documentation if needed.

amedina commented:
Consider using `Blocked plugins` in lieu of `Suppressed plugins`

westonruter commented:
> Consider using `Blocked plugins` in lieu of `Suppressed plugins`

I still prefer “suppressed” in favor of “blocked”. The term “block” is getting overly used in WordPress now, and I think “suppress” precisely depicts what is going on: “prevent the development, action, or expression of (a feeling, impulse, idea, etc.); restrain.”

> I am not too keen on putting the details of the offending plugin functionality on this screen. The amount of real estate is very limited and the details may not convey the right information.

I'm not sure. Where else would the information be displayed? The `details` elements are collapsed by default. How would someone discover understand what validation errors are being caused by the plugin, or to what extent a plugin is causing validation errors?

It's important for there to be as much information here as possible to help the user make a judgement as to whether the plugin can indeed be suppressed. This is why I added links off to the plugin homepage, the author name, and the plugin description.

In general, it could feel somewhat familiar to the plugins list table on the Plugins screen. That's kinda what I was aiming for.

> The description of the Suppressed plugins section is too verbose. We should simplify it and link to associated documentation if needed.

Agreed. It should be made more concise. But a benefit of it being in the plugin is that it will get translated. If we link off to docs on amp-wp.org then it won't get the benefit of translations (at least via translate.wordpress.org).

Nevertheless, it's important to note that this section only appears when there are plugins that are causing validation errors. Otherwise, the section is not shown at all. So in an ideal world, the appearance of this section should an be “exception”.

westonruter commented:
> In general, it could feel somewhat familiar to the plugins list table on the Plugins screen. That's kinda what I was aiming for.

In general, a table is able to present more more information in a structured way. My initial design of displaying the information in a linear list was inadequate for that purpose.

amedina commented:
> I still prefer “suppressed” in favor of “blocked”. The term “block” is getting overly used in WordPress now, and I think “suppress” precisely depicts what is going on: “prevent the development, action, or expression of (a feeling, impulse, idea, etc.); restrain.”

Makes sense. 
  
> I'm not sure. Where else would the information be displayed? The `details` elements are collapsed by default. How would someone discover understand what validation errors are being caused by the plugin, or to what extent a plugin is causing validation errors?

Is the goal to provide information to the user about the functionality of the plugin? And then the assumption is that users will use that information to get a better understanding of the potential issues being reported?

If that is the goal, my concern is that the functionality of a plugin is hard to capture in one sentence or so. For example, how to define the functionality of Jetpack? And having either a truncated description may not serve the purpose we want, and if we have a full description the UI can be cluttered. 
 
> It's important for there to be as much information here as possible to help the user make a judgement as to whether the plugin can indeed be suppressed. This is why I added links off to the plugin homepage, the author name, and the plugin description.

Agree. We should strike the right balance. See my concern above. 

> Nevertheless, it's important to note that this section only appears when there are plugins that are causing validation errors. Otherwise, the section is not shown at all. So in an ideal world, the appearance of this section should an be “exception”.

I agree with the goal. We may convey the same information more concisely.

westonruter committed:
Hide plugin author and description inside of details

westonruter commented:
> Is the goal to provide information to the user about the functionality of the plugin? And then the assumption is that users will use that information to get a better understanding of the potential issues being reported?

Yes.

> If that is the goal, my concern is that the functionality of a plugin is hard to capture in one sentence or so. For example, how to define the functionality of Jetpack? And having either a truncated description may not serve the purpose we want, and if we have a full description the UI can be cluttered.

The description being added here is the full description as provided by the plugin. If a user still doesn't understand what the plugin is doing from the description, they can click on the plugin title to visit the plugin homepage to get more information. You're right that the description can't encapsulate all of the information, but it can at least provide better context.

@amedina How about hiding the author name and description inside of `details` like so:

![image](https://user-images.githubusercontent.com/134745/83458706-cb494080-a417-11ea-9bbc-7ba8ab87220e.png)

This will hide the extra context unless the user needs it. With all details expanded:

![image](https://user-images.githubusercontent.com/134745/83458800-faf84880-a417-11ea-8cb3-07fb2549f8fa.png)

amedina commented:
> The description being added here is the full description as provided by the plugin. If a user still doesn't understand what the plugin is doing from the description, they can click on the plugin title to visit the plugin homepage to get more information. You're right that the description can't encapsulate all of the information, but it can at least provide better context.

Ok; that make sense to me. You are right. 
> 
> @amedina How about hiding the author name and description inside of `details` like so:

That would be good. Once the user suppresses a plugin they can collapse it.

westonruter commented:
Update based on our meeting: the list of validation errors should include the status of each, whether removed/kept and reviewed. Plugins that have all the validation errors reviewed would be less likely needing to be suppressed.

westonruter committed:
Move get_suppressible_plugins to PluginSuppression

westonruter committed:
Use array_intersect_key instead of wp_array_slice_assoc to preserver order

westonruter committed:
Update section to be named Plugin Suppression

westonruter committed:
Store URLs where errors occur and re-validate URL that has most errors present

westonruter committed:
Make storage of validated URLs dry

westonruter committed:
Only validate_suppressed_plugins when new suppressed plugins provided

westonruter committed:
Improve variable assignment formatting

westonruter committed:
Improve vertical alignment of suppressed plugins table

westonruter committed:
Improve formatting of plugin author and description

westonruter committed:
Capture the user who suppressed the plugin

westonruter committed:
Add validation error statuses to plugin suppression

westonruter commented:
In 5667e67e9a63edb25b8eb35388e9af3b7e83c6cc I've added _very_ abbreviated indications for the statuses of validation errors:

* Orange glow means unreviewed.
* Green bullet is removed invalid markup.
* Red bullet indicates kept invalid markup.

Disclaimer: I'm not a designer. But it feels pretty good, capturing the information concisely. Here's plugin expanded with validation errors in all 4 states:

![image](https://user-images.githubusercontent.com/134745/84553584-602a2480-acc9-11ea-9945-be930ed3de1e.png)

Each validation error link has a tooltip, in 4 possible values (for accessibility and discoverability):

1. Invalid markup causing the validation error is removed and reviewed. See all validated URL(s) with this validation error.
2. Invalid markup causing the validation error is kept and reviewed. See all validated URL(s) with this validation error.
3. Invalid markup causing the validation error is removed and unreviewed. See all validated URL(s) with this validation error.
4. Invalid markup causing the validation error is kept and unreviewed. See all validated URL(s) with this validation error.

schlessera commented:
⚠ ^^^ Force push happened here ^^^ ⚠

schlessera committed:
Remove empty file that was left over from rebase

schlessera committed:
Integrate plugin suppression with the services approach

schlessera committed:
Remove additional static access to plugin registry

schlessera committed:
Make PluginRegistry a shared instance

schlessera committed:
Fix tests involving plugin staleness

schlessera commented:
@googlebot I consent.

pierlon commented:
@googlebot I consent.

schlessera commented:
I don't think the Travis tests ran here.

schlessera committed:
Trigger Travis

schlessera commented:
Hmm, does Travis only run when the `develop` branch is targeted?

westonruter commented:
> Hmm, does Travis only run when the `develop` branch is targeted?

Yes. See this comment: https://github.com/ampproject/amp-wp/pull/4828#issuecomment-642447198

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/plugin-suppression

* 'develop' of github.com:ampproject/amp-wp:
  Fix PHP 5.6 back-compat code when ReflectionParameter has no class
  Fix Generic.Formatting.MultipleStatementAlignment
  Fix compat with PHP 5.6 where uniform variable syntax is not available
  Fix version command in lib/optimizer
  Use latest commit of PHPStan for now in lib/optimizer
  Use latest commit of PHPStan for now in lib/common
  Fix remaining PHPStan issues in main plugin
  Update Composer lock file
  Use latest commit of PHPStan for now
  Add PHPStan version output for CI debugging
  Remove exception for bug that was fixed upstream
  Remove autoload_files option

westonruter committed:
Fix phpcs issues

westonruter committed:
Finish refactoring factoring into service, fixing PHPStan analysis

westonruter committed:
Use bold for unreviewed validation errors, improve CSS formatting

westonruter committed:
Improve sort order of validation errors

westonruter committed:
Add tests for validating suppressed plugins

westonruter committed:
Remove redundant plugins cache

westonruter committed:
Add PluginRegistryTest and improve get_plugin_from_slug

westonruter commented:
I'm working on adding tests for the new functionality.

westonruter committed:
Obtain plugin dir from plugin_registry rather than WP_PLUGIN_DIR

schlessera committed:
Configure ServiceBasedPlugin for plugin use instead of theme use

westonruter committed:
Add majority of tests for PluginSuppression; account for the_widget()

westonruter committed:
Account for block editor not being present

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/plugin-suppression

* 'develop' of github.com:ampproject/amp-wp:
  Update dependency moment to v2.27.0

westonruter committed:
Account for widget_display_callback filter not being in the_widget() prior to WP 5.3

westonruter committed:
Add remaining tests for PluginSuppression

westonruter committed:
Update Suppressed Plugins description

westonruter committed:
Centralize option sanitization logic into Service

westonruter commented on includes/options/class-amp-options-manager.php:
```diff
@@ -286,6 +287,17 @@ public static function validate_options( $new_options ) {
+		 * @param array $options     Existing options with already-sanitized values for updating.
+		 * @param array $new_options Unsanitized options being submitted for updating.
+		 */
+		$options = apply_filters( 'amp_options_updating', $options, $new_options );
```
@schlessera I've added this to allow the option updating logic to be removed from the `AMP_Options_Manager` and over to the respective services.

schlessera replied:
I thought about turning the individual options into objects to inject as dependencies already. But the filter will do the trick for now.

BTW, this is what I referred to when I meant that we might need to add hooks so we can reverse the flow of the logic. So we seem to be on the same page wrt this.

schlessera commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -2135,6 +2124,45 @@ public static function validate_url( $url ) {
+			$args['invalid_url_post'] = $post;
+		}
+
+		$r = AMP_Validated_URL_Post_Type::store_validation_errors(
+			wp_list_pluck( $validity['results'], 'error' ),
+			$validity['url'],
+			$args
+		);
+		if ( $r instanceof WP_Error ) {
+			return $r;
+		}
+		$validity['post_id'] = $r;
```
Abbreviations like this should be avoided.
```suggestion
		$result = AMP_Validated_URL_Post_Type::store_validation_errors(
			wp_list_pluck( $validity['results'], 'error' ),
			$validity['url'],
			$args
		);
		if ( $result instanceof WP_Error ) {
			return $result;
		}
		$validity['post_id'] = $result;
```

schlessera commented on src/PluginRegistry.php:
```diff
@@ -0,0 +1,141 @@
+			}
+
+			$plugins[ $plugin_slug ] = array_merge(
+				[ 'File' => $plugin_file ], // Camel case is used for consistency with the other keys.
```
This is not camelCase, this is PascalCase (or StudlyCaps).
```suggestion
				[ 'File' => $plugin_file ], // PascalCase is used for consistency with the other keys.
```

schlessera reviewed (changes requested):
Looks good, only 2 nitpicks!

westonruter committed:
Improve variable name and comment

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>
